hadml/datamodules/components/herwig.py

```diff
@@ -661,113 +660,296 @@ def get_angles(four_vector):
+                        line, 
+                        cluster_rest_frame=cluster_rest_frame, 
+                        n_line=i if debug else None) 
+                    for i, line in enumerate(f) if line.strip()]
+            print()
 
             if self.pre_filter is not None:
                 data_list = [data for data in data_list if self.pre_filter(data)]
 
             if self.pre_transform is not None:
                 data_list = [self.pre_transform(data) for data in data_list]
+            all_data += data_list  
+
+        # Preparing diagrams depicting processed data distributions
+        if self.prepare_distribution_plots:
+            n_hadrons = [len(d.had_kin) for d in all_data]
+            hadron_energy = torch.cat([d.had_kin for d in all_data])[:, 0]
+            cluster_energy = torch.cat([d.x for d in all_data])[:, 0]
+            self._plot_dist(data=[n_hadrons, hadron_energy, cluster_energy],
+                    xlabels=["Number of hadrons", "Energy [GeV]", "Energy [GeV]"],
+                    ylabels=["Number of clusters", "Number of hadrons", "Number of clusters"],
+                    legend_labels=["Hadron Distribution", "Hadron Energy Distribution",
+                                "Cluster Energy Distribution"])
+        
+        # Saving the processed data as PyTorch binary files
+        if cluster_rest_frame == "both":
+            all_data_chosen_rest_frame = []
+            for d in all_data:
+                if self.cluster_rest_frame:
+                    had_kin = d.had_kin_rest_frame
+                else: 
+                    had_kin = d.had_kin
+                all_data_chosen_rest_frame += [
+                    Data(x=d.x, had_kin=had_kin, had_type_indices=d.had_type_indices,
+                    cluster_labels=d.cluster_labels)]
+            data, slices = self.collate(all_data_chosen_rest_frame)
+        else:
+            data, slices = self.collate(all_data)
+        torch.save((data, slices), self.processed_paths[0])
 
-            all_data += data_list
 
-        data, slices = self.collate(all_data)
-        torch.save((data, slices), self.processed_paths[0])
+    def _create_dict(self):
+        """ Create a new PID-to-MostCommonID map/dictionary and call the parsing method 
+        (self._create_data) to get all the hadron PIDs. """
+        all_data = []
 
-    def _create_data(self, line):
+        for raw_path in self.raw_file_list:
+            with open(raw_path) as f:
+                data_list = [self._create_data(line, create_dict=True, n_line=i) 
+                             for i, line in enumerate(f) if line.strip()]
+            print()
+            data_list = np.concatenate(data_list)
+            all_data.append(data_list)
+        
+        all_data = np.array(all_data).flatten()
+        count = Counter(all_data)
+        hadron_pids = list(map(lambda x: x[0], count.most_common()))
+        self.pids_to_ix = {pids: i for i, pids in enumerate(hadron_pids)}
```

Comment:
This name actually contains a typo which we never fixed for some reason - should be `pids_to_idx`

Reply:
Well, that was a mystery for me all the time but I thought the abbreviation had some sort of explanation. Anyway, I got it, so I will correct it as well.

Reply:
thanks